dora/underfs/hdfs/src/main/java/alluxio/underfs/hdfs/HdfsUnderFileSystem.java

```diff
@@ -751,15 +765,32 @@ public boolean supportsFlush() {
    *
    * @param path file or directory path
    * @param recursive whether to delete path recursively
+   * @param isDirectory whether path is a directory
    * @return true, if succeed
    */
-  private boolean delete(String path, boolean recursive) throws IOException {
+  private boolean delete(String path, boolean recursive, boolean isDirectory) throws IOException {
     IOException te = null;
     FileSystem hdfs = getFs();
     RetryPolicy retryPolicy = new CountingRetry(MAX_TRY);
     while (retryPolicy.attempt()) {
       try {
-        return hdfs.delete(new Path(path), recursive);
+        Path hdfsPath = new Path(path);
+        if (!mTrashEnable) {
+          return hdfs.delete(hdfsPath, recursive);
+        }
+        Trash trash = getTrash(hdfs);
+        // move to trash
+        if (isDirectory && !recursive && hdfs.listStatus(hdfsPath).length != 0) {
+          return false;
+        }
+        if (trash.moveToTrash(hdfsPath)) {
+          // moving file to trash succeeded.
+          return true;
+        } else {
+          // if failed to move this file to trash, delete it.
+          LOG.debug("Failed to move '{}' to trash. Now delete it.", path);
```

Comment:
Should we offer a configuration to control the direct throwing of exceptions, as some users might not want to transition from moving to the trash to directly deleting files?

Reply:
Different users may want different behaviors on such issue.
Some of them expect the files to be deleted.
Some others expect an exception/error.

Let's refine this requirement in future. But now I prefer to keep it simple.